cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -454,12 +531,49 @@ public void saveReport(HttpServiceRequest request, HttpServiceResponder responde
                            StandardCharsets.UTF_8);
     }
 
+    /**
+     * Use reportId and userName to construct the {@link ReportIdentifier}
+     * If instead shareId is provided, we decode the userName and reportId from shareId and construct the identifier
+     * and return that.
+     * @param reportId - id of the report, can be null if shareId is provided
+     * @param headers - request headers, used to retrieve the name of the owner who generated the report,
+     *                  if reportId is not null
+     * @param shareId - shareId if the report is a shared report, can be null if reportId and userName is provided
+     * @return {@link ReportIdentifier}
+     * @throws IllegalArgumentException if none of the reportId and shareId is provided or if both are provided
+     * @throws UnsupportedEncodingException if there are any error while decoding the shareId
+     */
+    private ReportIdentifier validateAndGetReportIdentifier(
+            @Nullable String reportId, Map<String, List<String>> headers,
+            @Nullable String shareId) throws IllegalArgumentException, UnsupportedEncodingException {
+      if (reportId == null && shareId == null) {
+        throw new IllegalArgumentException("Either reportId or sharedId must be provided, missing both");
+      }
+
+      if (reportId != null && shareId != null) {
+        throw new IllegalArgumentException("Only one of reportId or sharedId must be provided, provided both");
+      }
+      if (reportId != null) {
+        String userName = getUserName(headers);
+        return new ReportIdentifier(getUserName(headers), reportId);
+      } else {
+        // shareid is provided
+        return decodeShareId(shareId);
+      }
+    }
+
+    private Location getLocationFromReportIdentifier(ReportIdentifier reportIdentifier)
+            throws IOException {
+        return getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(reportIdentifier.getUserName()).
```

Comment:
looks like your IDE is using 4 spaces indentation instead of 2

Reply:
fixed it now, thanks